### PR TITLE
feat: min/max axis range for chart

### PR DESCRIPTION
**Pull request by nathaliabj**
### Pre requirements:
- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/lightdash/lightdash/blob/main/.github/CONTRIBUTING.md#sending-a-pull-request).

### Reference:
Closes: #84 

### Description:
This PR adds a min and max value for y-axis

### Preview:
Screenshot 2022-05-24 at 14 08 52
[Loom](https://www.loom.com/share/b7839a319aff4dae8a3308dc5a078781)

### Scope of the changes (select all that apply):
- [x] Frontend  
- [ ] Backend  
- [x] Common  
- [ ] E2E tests  
- [ ] Docs  
- [ ] CI/CD

### CODE REVIEW

File: packages/frontend/src/components/ChartConfigPanel/ChartConfigTabs.tsx [hunk 1]
```diff
@@ -32,19 +102,22 @@ const ChartConfigTabs: FC = () => {
...
-
     const [tab, setTab] = useState<string | number>('layout');
 
     const dimensionsInMetricQuery = explore
         ? getDimensions(explore).filter((field) =>
               resultsData?.metricQuery.dimensions.includes(fieldId(field)),
           )
         : [];
-
+    console.log({ resultsData, dirtyLayout });
```
Comment (ZeRego) [inline packages/frontend/src/components/ChartConfigPanel/ChartConfigTabs.tsx:120]: forgotten console log

File: packages/frontend/src/components/ChartConfigPanel/ChartConfigTabs.tsx [hunk 2]
```diff
@@ -15,12 +15,82 @@ import {
...
+        (axis === 'y'
+            ? dirtyEchartsConfig?.yAxis?.[index]?.max?.toString()
+            : dirtyEchartsConfig?.xAxis?.[index]?.max?.toString()) || undefined;
+
+    return (
+        <MinMaxContainer>
+            <Switch
+                name="auto-range"
+                checked={isAuto}
+                label={`Auto ${axis}-axis range`}
```
Comment (ZeRego) [inline packages/frontend/src/components/ChartConfigPanel/ChartConfigTabs.tsx:60]: This label will be inconsistent when axis are flipped

File: packages/frontend/src/components/ChartConfigPanel/ChartConfigTabs.tsx [hunk 3]
```diff
@@ -15,12 +15,82 @@ import {
...
 
+interface MinMaxProps {
+    minVal: (index: number, value: string | number | undefined) => void;
+    maxVal: (index: number, value: string | number | undefined) => void;
+    index: number;
+    axis: string;
+}
+
+const AxisMinMax: FC<MinMaxProps> = ({ minVal, maxVal, index, axis }) => {
+    const [isAuto, setIsAuto] = useState<boolean>(true);
```
Comment (ZeRego) [inline packages/frontend/src/components/ChartConfigPanel/ChartConfigTabs.tsx:40]: Fyi: we have a useToggle hook for situations like this that might simplify the code slightly.

The isAuto is `true` every time we load the component. Should the value depend if there is a min/max value and remove the min/max values when user clicks back to be auto ?

File: packages/frontend/src/components/ChartConfigPanel/ChartConfigTabs.tsx [hunk 4]
```diff
@@ -15,12 +15,82 @@ import {
...
 import GroupedSeriesConfiguration from './Series/GroupedSeriesConfiguration';
 
+interface MinMaxProps {
+    minVal: (index: number, value: string | number | undefined) => void;
+    maxVal: (index: number, value: string | number | undefined) => void;
+    index: number;
+    axis: string;
+}
+
+const AxisMinMax: FC<MinMaxProps> = ({ minVal, maxVal, index, axis }) => {
```
Comment (ZeRego) [inline packages/frontend/src/components/ChartConfigPanel/ChartConfigTabs.tsx:39]: What you think if this component was "dumber" ? it wouldn't depend on `useVisualizationContext` or have axis logic: 
```
interface MinMaxProps {
    label: string;
    min:  string | number | undefined;
    max:  string | number | undefined;
    setMin: (index: number, value: string | number | undefined) => void;
    setMax: (index: number, value: string | number | undefined) => void;
}
```
Use:

```
<AxisMinMax
         label={`Auto ${dirtyLayout?.flipAxes ? 'X' : 'Y'}-axis range (${dirtyLayout?.flipAxes ? 'bottom' : 'left'})`}
         min={dirtyEchartsConfig?.yAxis?.[0]?.min}
         max={dirtyEchartsConfig?.yAxis?.[0]?.max}
         setMin={setYMinValue}
         setMax={setYMaxValue}
/>
```

File: packages/frontend/src/components/ChartConfigPanel/ChartConfigTabs.tsx [hunk 5]
```diff
@@ -15,12 +15,82 @@ import {
...
+                    });
+                }}
+            />
+            {!isAuto && (
+                <MinMaxWrapper>
+                    <MinMaxInput label="Min">
+                        <InputGroup
+                            placeholder="Min"
+                            defaultValue={minDefaultValue}
+                            onBlur={(e) => minVal(index, e.currentTarget.value)}
```
Comment (ZeRego) [inline packages/frontend/src/components/ChartConfigPanel/ChartConfigTabs.tsx:77]: Isn't the value always string and eCharts supports a number as string "5" ? if so, should we only support min/max property as `string | undefined` ?

**Review by ZeRego — CHANGES_REQUESTED** [review]

File: packages/frontend/src/hooks/useEcharts.ts [hunk 6]
```diff
@@ -338,6 +338,12 @@ const getEchartAxis = ({
                       })
                     : xAxisConfiguration?.[0]?.name ||
                       (xAxisItem ? getItemLabel(xAxisItem) : undefined),
+                min:
+                    validCartesianConfig?.eChartsConfig?.xAxis?.[0]?.min ||
```
Comment (ZeRego) [inline packages/frontend/src/hooks/useEcharts.ts:342]: I believe we need to use `yAxisConfiguration` and `xAxisConfiguration` variables so it works when the axis are flipped.

File: packages/frontend/src/components/ChartConfigPanel/ChartConfigTabs.tsx (hunk 4 above)
Comment (ZeRego) [reply packages/frontend/src/components/ChartConfigPanel/ChartConfigTabs.tsx:39]: The AxisMinMax component shouldn't care about the index if is a ""dumb" component, that should be applied outside 

```
<AxisMinMax
         label={`Auto ${dirtyLayout?.flipAxes ? 'X' : 'Y'}-axis range (${dirtyLayout?.flipAxes ? 'bottom' : 'left'})`}
         min={dirtyEchartsConfig?.yAxis?.[0]?.min}
         max={dirtyEchartsConfig?.yAxis?.[0]?.max}
         setMin={(newValue: string | undefined ) => setYMinValue(0, newValue)}
         setMax={(newValue: string | undefined ) => setYMaxValue(0, newValue)}
/>
```

**Comment by hamzahc1** [comment]
@nathaliabj When I tried the preview env, the min-max switch + fields shows up for all types (not just numbers) but you can only input numbers right now. This means it doesn't work for things like dates, strings etc, in those cases should the ranges be disabled for now? 

Also - in this example the chart config showed me 2 x-axis configs... although actually there's only one on the chart. and manually changing the ranges didn't change the charts:
Screenshot 2022-05-25 at 10 53 59

**Comment by TuringLovesDeathMetal** [comment]
@nathaliabj just popping this here for transparency!

We said, priorities of things to change:

1. These `range` options should only appear when the values are _**numeric**_
2. There should be no `x-axis` option (only y-axis option). Except on horizontal bar charts (where there's an `x-axis` option, but no `y-axis` option)

Otherwise, I'm going to keep playing around with it and see what else I can spot 👀

**Comment by PriPatel** [comment]
> The right-axis range keeps popping up (even if there's no right-axis)
This is not a bug! We left it there because it was taking hours to fix. 
@ZeRego this is the thing we tried to address when working on chart config but its quite complex, correct?

**Comment by ZeRego** [comment]
> > The right-axis range keeps popping up (even if there's no right-axis)
> 
> This is not a bug! We left it there because it was taking hours to fix. @ZeRego this is the thing we tried to address when working on chart config but its quite complex, correct?

If we are talking about the second y axis configuration inputs, that is correct.

**Comment by TuringLovesDeathMetal** [comment]
> This is not a bug! We left it there because it was taking hours to fix.

hmmm...this definitely feels like a bug. (but it may be an intentional bug haha). 

@ZeRego is this impossible? Or just hard to do?

**Comment by TuringLovesDeathMetal** [comment]
@ZeRego the default range is always min = 0 (not min = min value). This was specified in the ticket, but I'm not sure if it's a massive change? (if yes, we can create a new ticket for it)

E.g. here, I would expect that the min value by default would be the min value in the chart (30), not 0

image

**Pull request merged**